lib/rmt/file_validator.rb

```diff
@@ -0,0 +1,59 @@
+require 'fileutils'
+require 'rmt/checksum_verifier'
+
+class RMT::FileValidator
+  class << self
+    def validate_local_file(file_reference, deep_verify:)
+      verify_file_on_disk(file_reference, deep_verify).tap do |valid_on_disk|
+        update_file_status_on_database(file_reference, valid_on_disk)
+      end
+    end
+
+    def find_valid_file_by_checksum(checksum, checksum_type, deep_verify:)
+      ::DownloadedFile
+        .where(checksum: checksum, checksum_type: checksum_type)
+        .map { |file| [file, validate_tracked_file_on_disk(file, deep_verify)] }
+        .find { |_, valid_on_disk| valid_on_disk }
+        &.first&.local_path
+    end
+
+    private
+
+    def validate_tracked_file_on_disk(file_record, deep_verify)
+      verify_file_on_disk(file_record, deep_verify).tap do |valid_on_disk|
+        file_record.destroy unless valid_on_disk
+      end
+    end
+
+    def verify_file_on_disk(file, deep_verify)
+      file_exist = File.exist?(file.local_path)
+      match_metadata = file_exist ? file_match_metadata?(file, deep_verify) : false
```

Comment:
Or you could do:

```ruby
match_metadata = (File.size(file.local_path) == file.size)
if deep_verify
  match_metadata &= RMT::ChecksumVerifier.match_checksum?(
    file.checksum_type, file.checksum, file.local_path
  )
end
```

And then you don't need those one-use one-line methods and you don't have to pass `deep_verify` two levels down into them.

Also `match_metadata` should be called either `matches_metadata` or better yet `is_valid_metadata`.

Reply:
I liked your suggestion, I was annoyed to pass `deep_verify` multiple levels down. Thanks :)

I implemented it in the last commit :heavy_check_mark: